crates/shared/src/price_estimation/trade_verifier/balance_overrides.rs

```diff
@@ -71,90 +186,99 @@ impl Strategy {
     }
 }
 
-impl ConfigurationBalanceOverrides {
-    pub fn new(config: HashMap<Address, Strategy>) -> Self {
-        Self(config)
-    }
-}
+type DetectorCache = Mutex<SizedCache<Address, Option<Strategy>>>;
 
-impl BalanceOverriding for ConfigurationBalanceOverrides {
-    fn state_override(&self, request: &BalanceOverrideRequest) -> Option<StateOverride> {
-        let strategy = self.0.get(&request.token)?;
-        let (key, value) = strategy.state_override(&request.holder, &request.amount);
-        Some(StateOverride {
-            state_diff: Some(hashmap! { key => value }),
-            ..Default::default()
-        })
-    }
+/// The default balance override provider.
+#[derive(Debug, Default)]
+pub struct BalanceOverrides {
+    /// The configured balance override strategies for tokens.
+    ///
+    /// These take priority over the auto-detection mechanism and are excluded
+    /// from the cache in order to prevent them from getting cleaned up by
+    /// the caching policy.
+    hardcoded: HashMap<Address, Strategy>,
+    /// The balance override detector and its cache. Set to `None` if
+    /// auto-detection is not enabled.
+    detector: Option<(Detector, DetectorCache)>,
 }
 
-impl Display for ConfigurationBalanceOverrides {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        let format_entry =
-            |f: &mut Formatter, (addr, strategy): (&Address, &Strategy)| match strategy {
-                Strategy::Mapping { slot } => write!(f, "{addr:?}@{slot}"),
-            };
+impl BalanceOverrides {
+    async fn cached_detection(&self, token: Address) -> Option<Strategy> {
+        let (detector, cache) = self.detector.as_ref()?;
+        tracing::debug!(?token, "attempting to auto-detect");
```

Comment:
This and the `"cache hit"` log will be super noisy.
I think having logs for errors and when the value actually gets cached is sufficient.